<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="IE=edge" >
  <title>java种Throwable、Error、Exception、RuntimeException 区别 联系 | 明动天下</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <meta name="description" content="Throwable 类是 Java 语言中所有错误或异常的超类。它的两个子类是Error和Exception。
Exception：1．可以是可被控制(checked) 或不可控制的(unchecked) 
2．表示一个由程序员导致的错误 
3．应该在应用程序级被处理
Error：1．总是不可控制的(unchecked) 
2．经常用来用于表示系统错误或低层资源的错误 
3．如何可能的话，应该在系">
<meta property="og:type" content="article">
<meta property="og:title" content="java种Throwable、Error、Exception、RuntimeException 区别 联系">
<meta property="og:url" content="http://yoursite.com/2017/03/23/java中的异常和错误/index.html">
<meta property="og:site_name" content="明动天下">
<meta property="og:description" content="Throwable 类是 Java 语言中所有错误或异常的超类。它的两个子类是Error和Exception。
Exception：1．可以是可被控制(checked) 或不可控制的(unchecked) 
2．表示一个由程序员导致的错误 
3．应该在应用程序级被处理
Error：1．总是不可控制的(unchecked) 
2．经常用来用于表示系统错误或低层资源的错误 
3．如何可能的话，应该在系">
<meta property="og:updated_time" content="2017-03-27T11:28:55.256Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="java种Throwable、Error、Exception、RuntimeException 区别 联系">
<meta name="twitter:description" content="Throwable 类是 Java 语言中所有错误或异常的超类。它的两个子类是Error和Exception。
Exception：1．可以是可被控制(checked) 或不可控制的(unchecked) 
2．表示一个由程序员导致的错误 
3．应该在应用程序级被处理
Error：1．总是不可控制的(unchecked) 
2．经常用来用于表示系统错误或低层资源的错误 
3．如何可能的话，应该在系">
  
    <link rel="alternative" href="/atom.xml" title="明动天下" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  <link rel="stylesheet" href="/css/style.css">
</head>

<body>
  <div id="container">
    <div class="left-col">
    <div class="overlay"></div>
<div class="intrude-less">
	<header id="header" class="inner">
		<a href="/" class="profilepic">
			
			<img lazy-src="/img/me.jpg" class="js-avatar">
			
		</a>

		<hgroup>
		  <h1 class="header-author"><a href="/">明动天下</a></h1>
		</hgroup>

		
		<p class="header-subtitle">提笔写诗   落笔代码</p>
		

		
			<div class="switch-btn">
				<div class="icon">
					<div class="icon-ctn">
						<div class="icon-wrap icon-house" data-idx="0">
							<div class="birdhouse"></div>
							<div class="birdhouse_holes"></div>
						</div>
						<div class="icon-wrap icon-ribbon hide" data-idx="1">
							<div class="ribbon"></div>
						</div>
						
						
					</div>
					
				</div>
				<div class="tips-box hide">
					<div class="tips-arrow"></div>
					<ul class="tips-inner">
						<li>菜单</li>
						<li>标签</li>
						
						
					</ul>
				</div>
			</div>
		

		<div class="switch-area">
			<div class="switch-wrap">
				<section class="switch-part switch-part1">
					<nav class="header-menu">
						<ul>
						
							<li><a href="/">主页</a></li>
				        
							<li><a href="/archives">概览</a></li>
				        
						</ul>
					</nav>
					<nav class="header-nav">
						<div class="social">
							
								<a class="github" target="_blank" href="https://github.com/wmingstar" title="github">github</a>
					        
								<a class="zhihu" target="_blank" href="https://www.zhihu.com/people/wmingstar" title="zhihu">zhihu</a>
					        
						</div>
					</nav>
				</section>
				
				
				<section class="switch-part switch-part2">
					<div class="widget tagcloud" id="js-tagcloud">
						<a href="/tags/elasticsearch/" style="font-size: 10px;">elasticsearch</a> <a href="/tags/github/" style="font-size: 11.25px;">github</a> <a href="/tags/java/" style="font-size: 20px;">java</a> <a href="/tags/linux/" style="font-size: 17.5px;">linux</a> <a href="/tags/python/" style="font-size: 16.25px;">python</a> <a href="/tags/spark/" style="font-size: 12.5px;">spark</a> <a href="/tags/分布式系统/" style="font-size: 10px;">分布式系统</a> <a href="/tags/剑指Offer/" style="font-size: 11.25px;">剑指Offer</a> <a href="/tags/操作系统/" style="font-size: 12.5px;">操作系统</a> <a href="/tags/数据库/" style="font-size: 15px;">数据库</a> <a href="/tags/数据结构/" style="font-size: 10px;">数据结构</a> <a href="/tags/笔试/" style="font-size: 10px;">笔试</a> <a href="/tags/算法/" style="font-size: 18.75px;">算法</a> <a href="/tags/面试/" style="font-size: 13.75px;">面试</a>
					</div>
				</section>
				
				
				

				
			</div>
		</div>
	</header>				
</div>

    </div>
    <div class="mid-col">
      <nav id="mobile-nav">
  	<div class="overlay">
  		<div class="slider-trigger"></div>
  		<h1 class="header-author js-mobile-header hide">明动天下</h1>
  	</div>
	<div class="intrude-less">
		<header id="header" class="inner">
			<div class="profilepic">
			
				<img lazy-src="/img/me.jpg" class="js-avatar">
			
			</div>
			<hgroup>
			  <h1 class="header-author">明动天下</h1>
			</hgroup>
			
			<p class="header-subtitle">提笔写诗   落笔代码</p>
			
			<nav class="header-menu">
				<ul>
				
					<li><a href="/">主页</a></li>
		        
					<li><a href="/archives">概览</a></li>
		        
		        <div class="clearfix"></div>
				</ul>
			</nav>
			<nav class="header-nav">
				<div class="social">
					
						<a class="github" target="_blank" href="https://github.com/wmingstar" title="github">github</a>
			        
						<a class="zhihu" target="_blank" href="https://www.zhihu.com/people/wmingstar" title="zhihu">zhihu</a>
			        
				</div>
			</nav>
		</header>				
	</div>
</nav>

      <div class="body-wrap"><article id="post-java中的异常和错误" class="article article-type-post" itemscope itemprop="blogPost">
  
    <div class="article-meta">
      <a href="/2017/03/23/java中的异常和错误/" class="article-date">
  	<time datetime="2017-03-23T00:22:49.000Z" itemprop="datePublished">2017-03-23</time>
</a>
    </div>
  
  <div class="article-inner">
    
      <input type="hidden" class="isFancy" />
    
    
      <header class="article-header">
        
  
    <h1 class="article-title" itemprop="name">
      java种Throwable、Error、Exception、RuntimeException 区别 联系
    </h1>
  

      </header>
      
      <div class="article-info article-info-post">
        
	<div class="article-tag tagcloud">
		<ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/java/">java</a></li></ul>
	</div>

        

        <div class="clearfix"></div>
      </div>
      
    
    <div class="article-entry" itemprop="articleBody">
      
        <p>Throwable 类是 Java 语言中所有错误或异常的超类。它的两个子类是Error和Exception。</p>
<h3 id="Exception："><a href="#Exception：" class="headerlink" title="Exception："></a>Exception：</h3><p>1．可以是可被控制(checked) 或不可控制的(unchecked) </p>
<p>2．表示一个由程序员导致的错误 </p>
<p>3．应该在应用程序级被处理</p>
<h3 id="Error："><a href="#Error：" class="headerlink" title="Error："></a>Error：</h3><p>1．总是不可控制的(unchecked) </p>
<p>2．经常用来用于表示系统错误或低层资源的错误 </p>
<p>3．如何可能的话，应该在系统级被捕捉</p>
<h3 id="Checked-exception-与-Unchecked-exception"><a href="#Checked-exception-与-Unchecked-exception" class="headerlink" title="Checked exception 与 Unchecked exception"></a>Checked exception 与 Unchecked exception</h3><h4 id="Checked-exception"><a href="#Checked-exception" class="headerlink" title="Checked exception"></a>Checked exception</h4><p>这类异常都是Exception的子类，用异常的向上抛出机制进行处理，假如子类可能<br>产生A异常，那么在父类中也必须throws A异常。可能导致的问题：代码效率低，耦合度过高。</p>
<h4 id="Unchecked-exception"><a href="#Unchecked-exception" class="headerlink" title="Unchecked exception"></a>Unchecked exception</h4><p>这类异常都是RuntimeException的子类，虽然RuntimeException同样也是Exception的子类，但是它们<br>是非凡的，它们不能通过client code来试图解决，所以称为Unchecked exception 。</p>
<h4 id="Checked-Exception与Runtime-Exception-的区别"><a href="#Checked-Exception与Runtime-Exception-的区别" class="headerlink" title="Checked Exception与Runtime Exception 的区别"></a>Checked Exception与Runtime Exception 的区别</h4><p>Java 提供了两种Exception 的模式，一种是执行的时候所产生的Exception (Runtime Exception)，<br>另外一种则是受控制的Exception (Checked Exception)。</p>
<p>所有的Checked Exception 均从java.lang.Exception 继承而来，而Runtime Exception<br>则继承java.lang.RuntimeException 或java.lang.Error (实际上java.lang.RuntimeException<br>的上一层也是java.lang.Exception)。</p>
<p>写代码时很可能会对选择某种形式的Exception 感到困扰，到底应该选择Runtime Exception<br>还是Checked Exception ？</p>
<p>其实，在运作上，可以通过Class 的Method 如何产生某个Exception以及某个程序如何<br>处理这个被产生来的Exception 来了解它们之间的差异，先建立一个Exception：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">public class CException extends Exception &#123;  </span><br><span class="line">    </span><br><span class="line">    public <span class="function"><span class="title">CException</span></span>() &#123;  </span><br><span class="line">    &#125;  </span><br><span class="line">  </span><br><span class="line">    public CException(String message) &#123;  </span><br><span class="line">        super(message);  </span><br><span class="line">    &#125;  </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<p>测试程序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">public class TestException &#123;  </span><br><span class="line">    public void method1() throws CException &#123;  </span><br><span class="line">        throw new CException(<span class="string">"Test Exception"</span>);  </span><br><span class="line">    &#125;  </span><br><span class="line">  </span><br><span class="line">    public void method2(String msg) &#123;  </span><br><span class="line">        <span class="keyword">if</span> (msg == null) &#123;  </span><br><span class="line">            throw new NullPointerException(<span class="string">"Message is null"</span>);  </span><br><span class="line">        &#125;  </span><br><span class="line">    &#125;  </span><br><span class="line">  </span><br><span class="line">    public void method3() throws CException &#123;  </span><br><span class="line">        method1();  </span><br><span class="line">    &#125;  </span><br><span class="line">  </span><br><span class="line">    // 以下省略  </span><br><span class="line">    // ...  </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>method1 和 method2 的程序代码内都会产生 Exception，<br>但method3 的程序代码中(大括号内)，并没产生Exception，但在method3 的定义中，暗示了这个 method<br>可能产生CException。</p>
<p>调用 method1() 的程序，由于 method1 会抛出异常 必须将method1() 包含在try 与catch 中，如：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">public class Runtest &#123;  </span><br><span class="line">    // ....  </span><br><span class="line">    public static void main(String argv[]) &#123;  </span><br><span class="line">        TestException te = new TestException();  </span><br><span class="line">        try &#123;  </span><br><span class="line">            te.method1();  </span><br><span class="line">        &#125; catch (CException ce) &#123;  </span><br><span class="line">            // ....  </span><br><span class="line">            ce.printStackTrace();  </span><br><span class="line">        &#125;  </span><br><span class="line">    &#125;  </span><br><span class="line">    // ...  </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<p>虽然包含在try 与catch 中，并不表示这段程序码一定会收到CException，<br>但它的用意在于提醒调用者，执行这个method可能产生的意外，而使用者也必须要能针对这个意外做出相对应<br>的处理方式。</p>
<p>当使用者呼叫 method2() 时，并不需要使用try 和catch 将程序码包起来，因为method2 的定义中，<br>并没有throws 任何的Exception ，如：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">public class Runtest  </span><br><span class="line">&#123;  </span><br><span class="line">// ....  </span><br><span class="line">public static void main(String argv[])  </span><br><span class="line">&#123;  </span><br><span class="line">  </span><br><span class="line"><span class="built_in">test</span>Exception te = new <span class="built_in">test</span>Exception();  </span><br><span class="line">  </span><br><span class="line">// 不会产生 Exception  </span><br><span class="line">te.method2(<span class="string">"Hello"</span>);  </span><br><span class="line">  </span><br><span class="line">// 会产生 Exception  </span><br><span class="line">te.method2(null);  </span><br><span class="line">&#125;  </span><br><span class="line">// ...  </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>若一个 method 内部抛出的是 RuntimeException，那定义这个 method 时，不需要明确 throws Exception</p>
<p>调用 method 时，若 method 明确  throws Exception，则需要将这个 method 的调用放在 try catch<br>中，对抛出的异常情况进行处理。</p>
<p>调用 method 时，若 method 没有 throws Exception，则不需要将这个 method 的调用放在 try catch<br>中。</p>

      
    </div>
    
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2017/03/25/http请求方法/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption"><</strong>
      <div class="article-nav-title">
        
          http请求方法
        
      </div>
    </a>
  
  
    <a href="/2017/03/18/笔试题解析1/" id="article-nav-older" class="article-nav-link-wrap">
      <div class="article-nav-title">笔试题解析1</div>
      <strong class="article-nav-caption">></strong>
    </a>
  
</nav>

  
</article>


<div class="share_jia">
	<!-- JiaThis Button BEGIN -->
	<div class="jiathis_style">
		<span class="jiathis_txt">分享到: &nbsp; </span>
		<a class="jiathis_button_facebook"></a> 
    <a class="jiathis_button_twitter"></a>
    <a class="jiathis_button_plus"></a> 
    <a class="jiathis_button_tsina"></a>
		<a class="jiathis_button_cqq"></a>
		<a class="jiathis_button_douban"></a>
		<a class="jiathis_button_weixin"></a>
		<a class="jiathis_button_tumblr"></a>
    <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jtico jtico_jiathis" target="_blank"></a>
	</div>
	<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js?uid=1405949716054953" charset="utf-8"></script>
	<!-- JiaThis Button END -->
</div>






<div class="duoshuo">
	<!-- 多说评论框 start -->
	<div class="ds-thread" data-thread-key="java中的异常和错误" data-title="java种Throwable、Error、Exception、RuntimeException 区别 联系" data-url="http://yoursite.com/2017/03/23/java中的异常和错误/"></div>
	<!-- 多说评论框 end -->
	<!-- 多说公共JS代码 start (一个网页只需插入一次) -->
	<script type="text/javascript">
	var duoshuoQuery = {short_name:"true"};
	(function() {
		var ds = document.createElement('script');
		ds.type = 'text/javascript';ds.async = true;
		ds.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') + '//static.duoshuo.com/embed.js';
		ds.charset = 'UTF-8';
		(document.getElementsByTagName('head')[0] 
		 || document.getElementsByTagName('body')[0]).appendChild(ds);
	})();
	</script>
	<!-- 多说公共JS代码 end -->
</div>




</div>
      <footer id="footer">
  <div class="outer">
    <div id="footer-info">
    	<div class="footer-left">
    		&copy; 2017 明动天下
    	</div>
      	<div class="footer-right">
      		<a href="http://hexo.io/" target="_blank">Hexo</a>  Theme <a href="https://github.com/litten/hexo-theme-yilia" target="_blank">Yilia</a> by Litten
      	</div>
    </div>
  </div>
</footer>
    </div>
    
  <link rel="stylesheet" href="/fancybox/jquery.fancybox.css">


<script>
	var yiliaConfig = {
		fancybox: true,
		mathjax: true,
		animate: true,
		isHome: false,
		isPost: true,
		isArchive: false,
		isTag: false,
		isCategory: false,
		open_in_new: false
	}
</script>
<script src="http://7.url.cn/edu/jslib/comb/require-2.1.6,jquery-1.9.1.min.js"></script>
<script src="/js/main.js"></script>






<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
        processEscapes: true,
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    }
});

MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for(i=0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';                 
    }       
});
</script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


  </div>
</body>
</html>